### ISSUE TITLE
scaling in gradient accumulation

### ISSUE THREAD

**Issue by gouthamk1998** [ultralytics/yolov5#464]
## ❔Is the accumulated gradient scaled according to the nominal batch size and the actual batch size??

I assume that the purpose of accumulate in train.py is to simulate the effect of training with batch size 64 while using a smaller batch size like 16. But as far as I understood, at every step after computing the loss, the code directly computes the backward pass without scaling the loss according to the accumulate size. For example, if our nominal batch size, nbs=64 and actual batch size, bs=16, then accumulate steps will be 4. Then at every step shouldn't we scale the computed loss by 4 to exactly simulate the effect of training with batch size 64.??

**Comment by glenn-jocher** [comment]
No, that’s incorrect. The gradient is accumulated correctly by the code.

**Comment by gouthamk1998** [comment]
Can you please provide a bit more explanation because the standard way of doing gradient accumulation in simpler application is like this where the loss is scaled at every step according to accumulate value
```
model.zero_grad()                                   # Reset gradients tensors
for i, (inputs, labels) in enumerate(training_set):
    predictions = model(inputs)                     # Forward pass
    loss = loss_function(predictions, labels)       # Compute loss function
    loss = loss / accumulation_steps                # Normalize our loss (if averaged)
    loss.backward()                                 # Backward pass
    if (i+1) % accumulation_steps == 0:             # Wait for several backward steps
        optimizer.step()                            # Now we can do an optimizer step
        model.zero_grad()                           # Reset gradients tensors
        if (i+1) % evaluation_steps == 0:           # Evaluate the model when we...
            evaluate_model()                        # ...have no gradients accumulated
```
Is Yolov5 following some other strategy so that this scaling is avoided?.

**Comment by HardLaugh** [comment]
default setting losses are mean about gt_match_anchor_num, nor batch_size,

**Comment by HardLaugh** [comment]
i check yolov5 multiple version, default setting losses are mean about gt_match_anchor_num nor batch_size, and then return loss * batch_size, so dont need divide accumulation_steps

**Comment by nannanmath** [comment]
@HardLaugh Hi, I feel if the loss is multiplied by batch size, the final loss will be not independent from different batch size? Actually I really do not know why the loss need to be scaled by batch size.
